### ISSUE TITLE
How to use LAL Python when installed via Alire?

### ISSUE THREAD

**Issue by Heziode** [AdaCore/libadalang#961]
The title speaks for itself, how can I use LAL in Python?

I does the following things:

```bash
alr get libadalang
cd libadalang
alr build
```

Now, I triying to follow [this tutorial on the official user doc](https://docs.adacore.com/live/wave/libadalang/html/libadalang_ug/examples/aggregate_projects.html), but I don't know where to start.

LAL is not reconnized:

```
> python lal_playground 
Traceback (most recent call last):
  File "lal_playground", line 18, in <module>
    import libadalang
ModuleNotFoundError: No module named 'libadalang'
```

I can't find anything in the doc to make it work…

I am working on an ubuntu Docker (VSCode), so I the dev environment are on Linux (the host is MacOS).

---

Furthere more, from the [doc (building)](https://docs.adacore.com/live/wave/libadalang/html/libadalang_ug/building.html#building-libadalang):

> 2. Building Libadalang
> First, note that if you have access to a pre-built Libadalang package, bundled with GNAT Pro if you have a GNAT Pro subscription (for community users, we have a Libadalang crate on [Alire](https://alire.ada.dev/crates/libadalang)), then you do not need to go through the following steps: just follow the instructions in the README file that comes with this pre-built package.

There is no `README` ship with Alire…

**Comment by thvnx** [comment]
Hi,

You don't have to build anything, I think you forgot to source the environment. I tried this on my side and it works:
```
$ alr get libadalang
$ cd libadalang_XXX
$ eval `alr printenv`
$ ./scripts/lal_playground test.adb
--
-- libadalang playground
--

The file(s) passed as argument have been put into the u variable, or units if
there are multiple.

Enjoy!

In [1]: 
```

You can also use `alr exec -- cmd` instead of the printenv command (it execute the command `cmd` in the local Alire environement):
```
$ alr exec -- ./scripts/lal_playground test.adb
--
-- libadalang playground
--

The file(s) passed as argument have been put into the u variable, or units if
there are multiple.

Enjoy!
```

**Comment by thvnx** [comment]
Hi @Heziode , I assume your problem is solved so I'm closing the issue. Feel free to reopen if you need more help with this.

**Comment by grochoge** [comment]
I'm running into a similar issue. It seems `alr build` is only building `libadalang.a`. I don't see any .pyd files and there's nothing to install IPython so I just get "ModuleNotFoundError: No module named 'IPython'" when trying to run `alr exec -- python .\scripts\lal_playground`

**Comment by thvnx** [comment]
Hi @grochoge , looks like my previous comment was not completely true. Can you try the following commands:
```shell
$ alr get libadalang
$ cd libadalang_XXX
# Build the relocatable (.so) version of libadalang
$ LIBRARY_TYPE=relocatable alr build
# Install the libadalang python package
$ cd python
$ pip install .
$ cd -
# Set env
$ eval `alr printenv`
$ export LD_LIBRARY_PATH=$PWD/lib/relocatable/dev/ 
# Run lal_playground
$ ./scripts/lal_playground test.adb
```
I'm not an Alire expert so there is maybe a better way to setup the build process and environment. I'll try to find it out, or to fix the Alire crate so that the python package is easier to use.

If you get the `ModuleNotFoundError: No module named 'IPython'` error, install it with `pip install ipython`

**Comment by NathanielPoteat** [comment]
I was running into a similar issue trying to use the Python API from the libadalang alire crate. First, is it possible that these instructions can be put somewhere more visible so other people don't have to dig to figure this out? Second, now that I have built the Python API I am running into a new issue:

```OSError: dlopen(libadalang.dylib, 0x0006): tried: 'libadalang.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OSlibadalang.dylib' (no such file), '/opt/homebrew/lib/libadalang.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/opt/homebrew/lib/libadalang.dylib' (no such file), '/usr/lib/libadalang.dylib' (no such file, not in dyld cache), 'libadalang.dylib' (no such file)```

I am running this on an M2 MacBook Pro, but I have also had issues getting past this point on an x64 AMD Linux machine as well:

```Traceback (most recent call last):
  File "/home/*********/repos/libadalang/scripts/lal_playground", line 15, in <module>
    import libadalang
  File "/home/*********/repos/libadalang/venv/lib/python3.9/site-packages/libadalang/__init__.py", line 84, in <module>
    _c_lib = ctypes.cdll.LoadLibrary(_c_lib_path)
  File "/usr/lib/python3.9/ctypes/__init__.py", line 452, in LoadLibrary
    return self._dlltype(name)
  File "/usr/lib/python3.9/ctypes/__init__.py", line 374, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: libadalang.so: cannot open shared object file: No such file or directory```

Have I missed any steps when building the libadalang alr crate that would not create the libadalang object file?

**Comment by thvnx** [comment]
Hi @NathanielPoteat , did you correctly set the `relocatable` build type when building the crate?

When running `LIBRARY_TYPE=relocatable alr build` you should find `libadalang.so` under `lib/relocatable/dev/`.

**Comment by NathanielPoteat** [comment]
Thank you, @thvnx. After running `LIBRARY_TYPE=relocatable alr build` I was able to find `libadalang.so` under `lib/relocatable/dev` on my Linux machine. However, now I'm running into another error with a `libgpr2.so`. Do I have to follow the same steps for libgpr2?
```Traceback (most recent call last):
  File "/home/npoteat3-temp/repos/libadalang/scripts/lal_playground", line 15, in <module>
    import libadalang
ImportError: libgpr2.so: cannot open shared object file: No such file or directory``` 

I would also like to note that the `libadalang.dylib` file was not created on my Mac M2 machine following the same instructions.

**Comment by thvnx** [comment]
Strange! Did you follow the steps I mentioned [here](https://github.com/AdaCore/libadalang/issues/961#issuecomment-1903947783)? Especially the `pip install .` command (it should install all you need to run the lal_playground). I've just checked on a linux machine and it works.

For the M2 machine you should try the same list of commands (replacing `LD_LIBRARY_PATH` by `DYLD_LIBRARY_PATH`). But I'm not sure that Alire and our tools support this architecture right now.

**Comment by NathanielPoteat** [comment]
Yes, I went back through and did all of the steps listed above again. It might be related to this error:
```Processing /home/npoteat3-temp/repos/libadalang/python
Building wheels for collected packages: Libadalang
  Building wheel for Libadalang (setup.py) ... error
  ERROR: Command errored out with exit status 1:
   command: /home/npoteat3-temp/repos/libadalang/venv/bin/python3.9 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-req-build-cnit3ygr/setup.py'"'"'; __file__='"'"'/tmp/pip-req-build-cnit3ygr/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-aml0d9mi
       cwd: /tmp/pip-req-build-cnit3ygr/
  Complete output (6 lines):
  usage: setup.py [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]
     or: setup.py --help [cmd1 cmd2 ...]
     or: setup.py --help-commands
     or: setup.py cmd --help

  error: invalid command 'bdist_wheel'
  ----------------------------------------
  ERROR: Failed building wheel for Libadalang
  Running setup.py clean for Libadalang
Failed to build Libadalang
Installing collected packages: Libadalang
  Attempting uninstall: Libadalang
    Found existing installation: Libadalang 0.1
    Uninstalling Libadalang-0.1:
      Successfully uninstalled Libadalang-0.1
    Running setup.py install for Libadalang ... done
Successfully installed Libadalang-0.1
```

Which I don't remember seeing the first time I did `pip install .` in the `python` directory.

**Comment by thvnx** [comment]
Looks like there is something wrong with your Python environment. Do you have the wheel package installed?

**Comment by NathanielPoteat** [comment]
It turns out that I did not have the wheel package installed on my Linux machine. Installing that solved the error when running `pip install .`, but I still run into the `ImportError: libgpr2.so: cannot open shared object file: No such file or directory` error when trying to run `lal_playground`. Should the `LIBRARY_TYPE=relocatable alr build` also ensure all dependency libraries are relocatable as well?

**Comment by NathanielPoteat** [comment]
When I run `ldd libadalang.so`, I see 

```libgpr2.so => /home/npoteat3-temp/repos/libadalang/lib/relocatable/dev/./../../../../..//.local/share/alire/builds/libgpr2_24.0.0_eda3c693/ee49417aad43e0ae9e1023ffb3696aa8fb6611fed1f20124c2ee879ea9ef9f99/.build/release/lib-relocatable/libgpr2.so (0x00007fe75fc40000)
```.

However, I still get the same error (`ImportError: libgpr2.so: cannot open shared object file: No such file or directory`) unless I copy the .so file to the same directory as the file I'm trying to use. After copying said libgpr2.so file, I then run into the same issue but for another .so file: 

```Traceback (most recent call last):
  File "/home/npoteat3-temp/repos/libadalang/scripts/lal_playground", line 15, in <module>
    import libadalang
ImportError: liblangkit_support.so: cannot open shared object file: No such file or directory
```.

**Comment by NathanielPoteat** [comment]
Even after copying all .so files into the same directory, I get another error:
```Traceback (most recent call last):
  File "/home/npoteat3-temp/repos/libadalang/scripts/lal_playground", line 15, in <module>
    import libadalang
ImportError: dynamic module does not define module export function (PyInit_libadalang)
```

**Comment by thvnx** [comment]
There is something wrong with your environment if you have to copy the `*.so`s. That's probably an issue on the Alire side at this point. You should check that `alr printenv` correctly prints the paths to the Alire local share builds and that the command is correctly evaluated.

**Comment by NathanielPoteat** [comment]
Executing `alr printenv` prints all the `.so` file locations as expected, but it still throws the same errors when I delete them from the directory.
